engine/consensus/approvals/assignment_collector_statemachine.go

```diff
@@ -0,0 +1,278 @@
+			// which might create a CPU spike and congestion on locking.
+			// Instead, we use a workerPool to buffer all approvals and process at most a certain
+			// number of them concurrently.
+			asm.workerPool.Submit(task)
+		}
+		return nil
+	}
+
+	return fmt.Errorf("cannot transition from %s to %s: %w", expectedCurrentStatus.String(), newStatus.String(), ErrInvalidCollectorStateTransition)
+}
+
+// caching2Orphaned ensures that the collector is currently in state `CachingApprovals`
+// and replaces it by a newly-created OrphanAssignmentCollector.
+// Returns:
+// * CachingAssignmentCollector as of before the update
+// * ErrDifferentCollectorState if the AssignmentCollector's state is _not_ `CachingApprovals`
+// * all other errors are unexpected and potential symptoms of internal bugs or state corruption (fatal)
+func (asm *AssignmentCollectorStateMachine) caching2Orphaned() (*CachingAssignmentCollector, error) {
+	asm.Lock()
+	defer asm.Unlock()
+	cachingCollector, ok := asm.atomicLoadCollector().(*CachingAssignmentCollector)
+	if !ok {
+		return nil, fmt.Errorf("collectors current state is %s: %w",
+			cachingCollector.ProcessingStatus().String(), ErrDifferentCollectorState)
+	}
+	asm.collector.Store(&atomicValueWrapper{collector: NewOrphanAssignmentCollector(asm.AssignmentCollectorBase)})
+	return cachingCollector, nil
+}
+
+// verifying2Orphaned ensures that the collector is currently in state `VerifyingApprovals`
+// and replaces it by a newly-created OrphanAssignmentCollector.
+// Returns:
+// * VerifyingAssignmentCollector as of before the update
+// * ErrDifferentCollectorState if the AssignmentCollector's state is _not_ `VerifyingApprovals`
+// * all other errors are unexpected and potential symptoms of internal bugs or state corruption (fatal)
+func (asm *AssignmentCollectorStateMachine) verifying2Orphaned() (*VerifyingAssignmentCollector, error) {
+	asm.Lock()
+	defer asm.Unlock()
+	verifyingCollector, ok := asm.atomicLoadCollector().(*VerifyingAssignmentCollector)
+	if !ok {
+		return nil, fmt.Errorf("collectors current state is %s: %w",
+			verifyingCollector.ProcessingStatus().String(), ErrDifferentCollectorState)
+	}
+	asm.collector.Store(&atomicValueWrapper{collector: NewOrphanAssignmentCollector(asm.AssignmentCollectorBase)})
+	return verifyingCollector, nil
+}
+
+// caching2Verifying ensures that the collector is currently in state `CachingApprovals`
+// and replaces it by a newly-created VerifyingAssignmentCollector.
+// Returns:
+// * CachingAssignmentCollector as of before the update
+// * ErrDifferentCollectorState if the AssignmentCollector's state is _not_ `CachingApprovals`
+// * all other errors are unexpected and potential symptoms of internal bugs or state corruption (fatal)
+func (asm *AssignmentCollectorStateMachine) caching2Verifying() (*CachingAssignmentCollector, error) {
+	asm.Lock()
+	defer asm.Unlock()
+	cachingCollector, ok := asm.atomicLoadCollector().(*CachingAssignmentCollector)
```

Comment:
you are right, it may happen. I will add a test for this case.

Reply:
Actually it can't in our scenario, it may happen theoretically but caller of `ChangeProcessingStatus` mutually excludes at same lock so only one goroutine will change processing status at time. In other words in current implementation race cannot happen.

Reply:
I am so confused :) We currently have a check for cast but still I have added a test for it.